cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java

```diff
@@ -27,11 +28,20 @@
   @Override
   public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set<String> excludedTopics)
       throws KafkaCruiseControlException {
+    // First move the replica on the demoted brokers to the end of the replica list.
+    // If all the replicas are demoted, no change is made to the leader.
+    for (Broker b : clusterModel.demotedBrokers()) {
+      for (Replica r : b.replicas()) {
+        Partition p = clusterModel.partition(r.topicPartition());
+        p.moveReplicaToEnd(r);
+      }
+    }
     // Ignore the excluded topics because this goal does not move partitions.
     for (List<Partition> partitions : clusterModel.getPartitionsByTopic().values()) {
       for (Partition p : partitions) {
         for (Replica r : p.replicas()) {
-          // Iterate over the replicas and ensure the leader is set to the first alive replica.
+          // Iterate over the replicas and ensure the leader is set to the first alive replica and not
+          // demoted partition.
```

Comment:
If the leadership in a demoted or dead broker cannot be moved either because the leader is the only alive replica of the partition or the replication factor was `1`, cruise control silently ignores it.

I believe depending on the request type (Just a `PLE` or `demote`) we may want to log a message to indicate such a reason or throw an exception.